scripts/cleanup_datasets/pgcleanup.py

```diff
@@ -844,17 +850,66 @@ class PurgeDeletedHDAs(PurgesHDAs, RemovesMetadataFiles, RequiresDiskUsageRecalc
     )
 
 
+class PurgeOldHDAs(PurgesHDAs, RemovesMetadataFiles, RequiresDiskUsageRecalculation, Action):
+    """
+    - Mark purged all HistoryDatasetAssociations that are older than the specified number of days.
+    - Mark deleted all MetadataFiles whose hda_id is purged in this step.
+    - Mark deleted all ImplicitlyConvertedDatasetAssociations whose hda_parent_id is purged in this
+      step.
+    - Mark purged all HistoryDatasetAssociations for which an ImplicitlyConvertedDatasetAssociation
+      with matching hda_id is deleted in this step.
+    """
+
+    force_retry_sql = " AND NOT history_dataset_association.purged"
+    _action_sql = """
+        WITH purged_hda_ids
+          AS (     UPDATE history_dataset_association
+                      SET purged = true, deleted = true{update_time_sql}
+                    FROM dataset
+                    WHERE dataset.create_time < (NOW() AT TIME ZONE 'utc' - interval '%(days)s days')
```

Comment:
If I'm following the logic here of using dataset rather than HDA, this ensures that you can't just copy (or create new HDAs) in order to escape scratch history cleanup?

Reply:
That isn't why per se... I think the logic is just that it is simpler to find and looked more like the existing code in here... if we did it on HDA creation time - which one would we use? Happy to refine the behavior though if requested. One can imagine policies that could cut either way and be completely reasonable.